aten/src/ATen/native/cuda/TensorTransformations.cu

```diff
@@ -126,4 +126,65 @@ Tensor flip_cuda(const Tensor& self, IntList dims) {
   return out_tensor;
 }
 
+template <typename scalar_t>
+__global__
+void roll_cuda_kernel(scalar_t* in_tensor, scalar_t* out_tensor, int64_t N,
+                      int64_t roll_dim, int64_t shift, int64_t start,
+                      int64_t size, int64_t stride, int64_t total_dims) {
+  int64_t linear_index = blockIdx.x * blockDim.x + threadIdx.x;
+  if (linear_index >= N) {
+    return;
+  }
+  // roll dim idx is the index of linear_index along the rolling dimension.
+  int64_t roll_dim_idx = linear_index % (stride * size) / stride;
+  // index into the source data to find appropriate value.
+  int64_t source_idx = 0;
+  if( roll_dim_idx >= (size - start) ) {
+    source_idx = linear_index - (shift * stride);
+  } else {
+    source_idx = linear_index + (start * stride);
+  }
+  out_tensor[linear_index] = in_tensor[source_idx];
+}
+
+// Roll a tensor along a dimension
+Tensor roll_cuda(const Tensor& self, IntList shifts, IntList dims) {
+  // todo: support rolling along no or multiple dimensions as in numpy.roll.
+  AT_CHECK(dims.size() == 1, "only single dimension roll currently supported");
+  AT_CHECK(shifts.size() == dims.size(), "shifts and dimensions must align");
+
+  auto in_tensor = self;
+  if( ! self.is_contiguous() ) {
+    in_tensor = self.contiguous();
+  }
+  auto out_tensor = at::empty_like(in_tensor);
+  if (out_tensor.numel() == 0) {
+    return out_tensor;
+  }
+  const int64_t N = self.numel();
+  const int64_t dim = dims[0];
+  const int64_t size = self.size(dim);
+  int64_t start = (size - shifts[0]) % size;
+  // Behavior of % is different in C++ vs Python for negative numbers. This
+  // corrects the difference.
+  if( start < 0 ) start = start + size;
+
+  const int64_t block_size = 512;
```

Comment:
any reason you didn't use getApplyBlock() and getApplyGrid?

Reply:
I was copying from flip_cuda(), I actually meant to ask about this because I didn't like it. will update.